src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java

```diff
@@ -84,6 +110,52 @@ private void onInit(Level levelIn, ChunkPos pos, ChunkBiomeContainer chunkBiomeC
         if (!(biomes instanceof ColumnBiomeContainer)) {
             this.biomes = new ColumnBiomeContainer(levelIn.registryAccess().registryOrThrow(Registry.BIOME_REGISTRY), levelIn, levelIn);
         }
+
+        if (levelIn.isClientSide) {
+            lightHeightmap = new ClientLightSurfaceTracker(this);
+        } else {
+            lightHeightmap = new LightSurfaceTrackerWrapper(this);
+        }
+        // TODO might want 4 columns that share the same BigCubes to have a reference to the same CubeMap?
+        cubeMap = new CubeMap();
+    }
+
+    @Inject(
+            method = "<init>(Lnet/minecraft/server/level/ServerLevel;Lnet/minecraft/world/level/chunk/ProtoChunk;Ljava/util/function/Consumer;)V",
+            at = @At("RETURN")
+    )
+    private void onInitFromProtoChunk(ServerLevel serverLevel, ProtoChunk protoChunk, Consumer<LevelChunk> consumer, CallbackInfo ci) {
+        if (!this.isCubic()) {
+            return;
+        }
+        lightHeightmap = ((LightHeightmapGetter) protoChunk).getLightHeightmap();
+        cubeMap = ((CubeMapGetter) protoChunk).getCubeMap();
+    }
+
+    @Inject(
+            method = "setBlockState(Lnet/minecraft/core/BlockPos;Lnet/minecraft/world/level/block/state/BlockState;Z)Lnet/minecraft/world/level/block/state/BlockState;",
+            at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/levelgen/Heightmap;update(IIILnet/minecraft/world/level/block/state/BlockState;)Z", ordinal = 0)
+    )
+    private void onSetBlock(BlockPos pos, BlockState state, boolean moved, CallbackInfoReturnable<BlockState> cir) {
+        if (!this.isCubic()) {
+            return;
+        }
+        // TODO client side light heightmap stuff
```

Comment:
Still not sure what needs to be done with the client light heightmap, in terms of updating when there are client block updates vs. receiving updates from the server. Doesn't really matter for this PR, since we just want server-side light working for now. I'll make the comment clearer about what might need to be done though.